### ISSUE TITLE
NUnit report incomplete when used with xdist

### ISSUE THREAD

**Issue by erigusaab** [pytest-dev/pytest-nunit#40]
I run some fairly "heavy" tests that really need parallel execution to speed up. I have been using pytest-xdist and --junit-xml for the report which I then import to our on-premises azure devops server.

Now, I would like to add some attachments so I want to switch to NUnit output but unfortunately I only get a summary in the XML file and no test case details reported. 

I can replicate this with a very minimal test case:

```
import pytest

def test_nunitxdist():
    assert 1 == 1
```

If I run `pytest -v --nunit-xml=nu.xml` I get the full report, but with `pytest -v --nunit-xml=nux2.xml -n2` I get a file with no "test-suite" tag at all

This is with:

```
platform linux -- Python 3.5.2, pytest-5.2.2, py-1.8.0, pluggy-0.13.0 -- /usr/bin/python3
plugins: nunit-0.5.2, forked-1.1.3, xdist-1.30.0
```

**Comment by erigusaab** [comment]
There appears to be a regression in v0.5.x, v0.4.0 is working partially - attachments are still missing in v0.4.0 but the test results are reported.

**Comment by tonybaloney** [comment]
Is it related to this comment? https://github.com/pytest-dev/pytest-nunit#x-dist

**Comment by ivirshup** [comment]
Any updates on this? 

I'm seeing reduced reports with `0.6.0`, but noticed some commits related to `xdist`.

**Comment by webshoptest** [comment]
I use pytest-nunit 0.6.0 and I get the error when trying to 'publish test results xml':

> No Result Found to Publish 'D:\a\1\s\nunit\test-results.xml'.

I even tried pytest-nunit 0.40 as suggested but it didn't work.
Without xdist is just works fine.

**Comment by amezin** [comment]
It happens because `pytest_collection_modifyitems()` hook runs in worker processes (during test collection), but other hooks are run in the controller process. So `node_to_module_map` ends up empty.

Why is a more complex logic is necessary for NUnit, compared to JUnit? Pytest's JUnit reporter outputs one flat test suite. As a user, I'd expect pytest-nunit to output a report with a similar structure. At least by default.

And for "structured" report, there are at least two alternative approaches, that will both work with xdist:
1) Use `[TestReport.fspath](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.TestReport.fspath)` as the module name
2) Parse `nodeid` to get parent node id. It isn't as fragile as it might sound - [pytest does it itself](https://github.com/pytest-dev/pytest/blob/5800d39053af398eebcefdb668e767acb7944f01/src/_pytest/nodes.py#L51-L93), when [loading](https://github.com/pytest-dev/pytest/blob/5800d39053af398eebcefdb668e767acb7944f01/src/_pytest/fixtures.py#L1463) [fixtures](https://github.com/pytest-dev/pytest/blob/5800d39053af398eebcefdb668e767acb7944f01/src/_pytest/fixtures.py#L1652)

**Comment by amezin** [comment]
By the way: pytest-nunit's code calls it "grouping by modules", but in my case it ends up being "grouping by classes"

**Comment by tonybaloney** [comment]
I think the bug might be here then based on what you've said:


https://github.com/pytest-dev/pytest-nunit/blob/master/pytest_nunit/plugin.py#L295-L299

**Comment by amezin** [comment]
No, those lines are unrelated.

`pytest_collection_modifyitems()` is the problem. It's called in a worker process, not the main process where the report is generated. And it is how xdist is designed. This hook can't be used the way it's used in the plugin when xdist is active.

**Comment by tonybaloney** [comment]
Pushing a fix in #57 

This will map nodes to test-suites in `pytest_sessionfinish` _if_ the node map is empty (e.g. in xdist), this way the behaviour with previous versions doesn't change.
